File: ch07/VbServerControl1/Listing07-42.vb
Imports System.Web.UI.Design
Imports System.ComponentModel.Design
Imports System.ComponentModel
Imports System.Web.UI

Public Class Listing0742Designer
    Inherits ControlDesigner

    Private _actionLists As DesignerActionListCollection

    Public Overrides ReadOnly Property ActionLists() _
            As DesignerActionListCollection
        Get
            If IsNothing(_actionLists) Then
                _actionLists = New DesignerActionListCollection()
                _actionLists.AddRange(MyBase.ActionLists)
                _actionLists.Add(New ServerControl44ControlList(Me))
            End If
            Return _actionLists
        End Get
    End Property

    Private NotInheritable Class ServerControl44ControlList
        Inherits DesignerActionList

        Public Sub New(ByVal c As Listing0742Designer)
            MyBase.New(c.Component)
        End Sub

        Public Overrides Function GetSortedActionItems() _
                As DesignerActionItemCollection

            Dim c As New DesignerActionItemCollection()
            c.Add(New DesignerActionTextItem("Text Action Item",
                                             "Custom Category"))
            Return c
        End Function
    End Class
End Class

<DefaultProperty("Text"),
DesignerAttribute(GetType(Listing0742Designer)),
ToolboxData("<{0}:ServerControl44 runat=server></{0}:ServerControl44>")>
Public Class Listing0742
    Inherits System.Web.UI.WebControls.WebControl

    <Bindable(True), Category("Appearance"), DefaultValue(""), Localizable(True)>
    Property Text() As String
        Get
            Dim s As String = CStr(ViewState("Text"))
            If s Is Nothing Then
                Return "[" + Me.ID + "]"
            Else
                Return s
            End If
        End Get

        Set(ByVal Value As String)
            ViewState("Text") = Value
        End Set
    End Property

    Protected Overrides Sub RenderContents(ByVal output As HtmlTextWriter)
        output.Write(Text)
    End Sub
End Class

File: ch07/CSharpServerControl1/Listing07-42.cs
using System;
using System.ComponentModel;
using System.ComponentModel.Design;
using System.Web.UI;
using System.Web.UI.Design;
using System.Web.UI.WebControls;

namespace CSharpServerControl1
{
    public class Listing0742Designer : ControlDesigner
    {
        private DesignerActionListCollection _actionLists = null;

        public override DesignerActionListCollection ActionLists
        {
            get
            {
                if (_actionLists == null)
                {
                    _actionLists = new DesignerActionListCollection();
                    _actionLists.AddRange(base.ActionLists);
                    _actionLists.Add(new ServerControl44ControlList(this));
                }
                return _actionLists;
            }
        }

        private sealed class ServerControl44ControlList :
                                DesignerActionList
        {
            public ServerControl44ControlList(ControlDesigner c)
                : base(c.Component)
            {
            }

            public override DesignerActionItemCollection
                GetSortedActionItems()
            {
                DesignerActionItemCollection c =
                    new DesignerActionItemCollection();
                c.Add(new DesignerActionTextItem("Text Action Item",
                                                    "Custom Category"));
                return c;
            }
        }
    }

    [DesignerAttribute(typeof(Listing0742Designer))]
    [ToolboxData("<{0}:Listing0742 runat=server></{0}:Listing0742>")]
    public class Listing0742 : WebControl
    {
        [Bindable(true)]
        [Category("Appearance")]
        [DefaultValue("")]
        [Localizable(true)]
        public string Text
        {
            get
            {
                String s = (String)ViewState["Text"];
                return ((s == null) ? "[" + this.ID + "]" : s);
            }
            set
            {
                ViewState["Text"] = value;
            }
        }

        protected override void RenderContents(HtmlTextWriter output)
        {
            output.Write(Text);
        }
    }
}